agent/exec/containerd/adapter.go

```diff
@@ -0,0 +1,682 @@
+			}
+
+			s.Mounts = append(s.Mounts, mt)
+			continue
+		}
+	}
+
+	for v := range volumes {
+		mt, err := c.makeAnonVolume(ctx, v)
+		if err != nil {
+			return err
+		}
+
+		s.Mounts = append(s.Mounts, mt)
+	}
+	return nil
+}
+
+func (c *containerAdapter) spec(ctx context.Context, config *ocispec.ImageConfig, rootfs string) (*specs.Spec, error) {
+	caps := []string{
+		"CAP_CHOWN",
+		"CAP_DAC_OVERRIDE",
+		"CAP_FSETID",
+		"CAP_FOWNER",
+		"CAP_MKNOD",
+		"CAP_NET_RAW",
+		"CAP_SETGID",
+		"CAP_SETUID",
+		"CAP_SETFCAP",
+		"CAP_SETPCAP",
+		"CAP_NET_BIND_SERVICE",
+		"CAP_SYS_CHROOT",
+		"CAP_KILL",
+		"CAP_AUDIT_WRITE",
+	}
+
+	// Need github.com/docker/docker/oci.DefaultSpec()
+	spec := specs.Spec{
+		Version: "1.0.0-rc2-dev",
+		Root: specs.Root{
+			Path: rootfs,
+		},
+		Mounts: []specs.Mount{
+			{
+				Destination: "/proc",
+				Type:        "proc",
+				Source:      "proc",
+				Options:     []string{"nosuid", "noexec", "nodev"},
+			},
+			{
+				Destination: "/dev",
+				Type:        "tmpfs",
+				Source:      "tmpfs",
+				Options:     []string{"nosuid", "strictatime", "mode=755"},
+			},
+			{
+				Destination: "/dev/pts",
+				Type:        "devpts",
+				Source:      "devpts",
+				Options:     []string{"nosuid", "noexec", "newinstance", "ptmxmode=0666", "mode=0620", "gid=5"},
+			},
+			{
+				Destination: "/sys",
+				Type:        "sysfs",
+				Source:      "sysfs",
+				Options:     []string{"nosuid", "noexec", "nodev", "ro"},
+			},
+			{
+				Destination: "/sys/fs/cgroup",
+				Type:        "cgroup",
+				Source:      "cgroup",
+				Options:     []string{"ro", "nosuid", "noexec", "nodev"},
+			},
+			{
+				Destination: "/dev/mqueue",
+				Type:        "mqueue",
+				Source:      "mqueue",
+				Options:     []string{"nosuid", "noexec", "nodev"},
+			},
+		},
+		Process: specs.Process{
+			Cwd: "/",
+			Capabilities: &specs.LinuxCapabilities{
+				Bounding:    caps,
+				Effective:   caps,
+				Inheritable: caps,
+				Permitted:   caps,
+				Ambient:     caps,
+			},
+			NoNewPrivileges: true,
+			Terminal:        false,
+		},
+		Linux: &specs.Linux{
+			Namespaces: []specs.LinuxNamespace{
+				{Type: "mount"},
+				{Type: "network"},
+				{Type: "uts"},
+				{Type: "pid"},
+				{Type: "ipc"},
+			},
+		},
+	}
+
+	spec.Platform = specs.Platform{
+		OS:   runtime.GOOS,
+		Arch: runtime.GOARCH,
+	}
+	if config.WorkingDir != "" {
+		spec.Process.Cwd = config.WorkingDir
+	}
+	spec.Process.Env = config.Env
+
+	var args []string
+	if len(c.container.Args) > 0 {
+		args = c.container.Args
+	} else {
+		args = config.Cmd
+	}
+
+	if len(c.container.Command) > 0 {
+		spec.Process.Args = append(c.container.Command, args...)
+	} else {
+		spec.Process.Args = append(config.Entrypoint, args...)
+	}
+
+	log.G(ctx).Debugf("Process args: %v", spec.Process.Args)
+	if err := c.setMounts(ctx, &spec, c.container.Mounts, config.Volumes); err != nil {
+		return nil, errors.Wrap(err, "Failed to set mounts")
+	}
+	sort.Sort(mounts(spec.Mounts))
```

Comment:
Yes, the `mounts` type (see end of this file) is sorting them by path length, so that mounts nearest the root of the tree always happen before ones further out i.e. you never "shadow" a mount completely. (I believe docker engine has some similar code in it somewhere, but it's been a while and I don't remember where exactly).